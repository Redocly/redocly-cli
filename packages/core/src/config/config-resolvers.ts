import * as path from 'path';
import { blue, red } from 'colorette';
import { isAbsoluteUrl } from '../ref-utils';
import { BaseResolver } from '../resolve';
import { defaultPlugin } from './builtIn';
import {
  getResolveConfig,
  getUniquePlugins,
  mergeExtends,
  parsePresetName,
  prefixRules,
  transformConfig,
} from './utils';
import type {
  LintRawConfig,
  Plugin,
  RawConfig,
  ResolvedApi,
  ResolvedLintConfig,
  RuleConfig,
} from './types';
import { isNotString, isString, notUndefined, parseYaml } from '../utils';
import { Config } from './config';

export async function resolveConfig(rawConfig: RawConfig, configPath?: string) {
  if (rawConfig.lint?.extends?.some(isNotString)) {
    throw new Error(
      `Error configuration format not detected in extends value must contain strings`,
    );
  }

  const resolver = new BaseResolver(getResolveConfig(rawConfig.resolve));
  const configExtends = rawConfig?.lint?.extends ?? ['recommended'];
  const recommendedFallback = !rawConfig?.lint?.extends;
  const lintConfig = {
    ...rawConfig?.lint,
    extends: configExtends,
    recommendedFallback,
  };

  const apis = await resolveApis({
    rawConfig: {
      ...rawConfig,
      lint: lintConfig,
    },
    configPath,
    resolver,
  });

  const lint = await resolveLint({
    lintConfig,
    configPath,
    resolver,
  });

  return new Config(
    {
      ...rawConfig,
      apis,
      lint,
    },
    configPath,
  );
}

export function resolvePlugins(
  plugins: (string | Plugin)[] | null,
  configPath: string = '',
): Plugin[] {
  if (!plugins) return [];

  // @ts-ignore
  const requireFunc = typeof __webpack_require__ === 'function' ? __non_webpack_require__ : require;

  const seenPluginIds = new Map<string, string>();

  return plugins
    .map((p) => {
      if (isString(p) && isAbsoluteUrl(p)) {
        throw new Error(red(`We don't support remote plugins yet.`));
      }

      // TODO: resolve npm packages similar to eslint
      const pluginModule = isString(p)
        ? (requireFunc(path.resolve(path.dirname(configPath), p)) as Plugin)
        : p;

      const id = pluginModule.id;
      if (typeof id !== 'string') {
        throw new Error(red(`Plugin must define \`id\` property in ${blue(p.toString())}.`));
      }

      if (seenPluginIds.has(id)) {
        const pluginPath = seenPluginIds.get(id)!;
        throw new Error(
          red(
            `Plugin "id" must be unique. Plugin ${blue(p.toString())} uses id "${blue(
              id,
            )}" already seen in ${blue(pluginPath)}`,
          ),
        );
      }

      seenPluginIds.set(id, p.toString());

      const plugin: Plugin = {
        id,
        ...(pluginModule.configs ? { configs: pluginModule.configs } : {}),
        ...(pluginModule.typeExtension ? { typeExtension: pluginModule.typeExtension } : {}),
      };

      if (pluginModule.rules) {
        if (!pluginModule.rules.oas3 && !pluginModule.rules.oas2) {
          throw new Error(`Plugin rules must have \`oas3\` or \`oas2\` rules "${p}.`);
        }
        plugin.rules = {};
        if (pluginModule.rules.oas3) {
          plugin.rules.oas3 = prefixRules(pluginModule.rules.oas3, id);
        }
        if (pluginModule.rules.oas2) {
          plugin.rules.oas2 = prefixRules(pluginModule.rules.oas2, id);
        }
      }
      if (pluginModule.preprocessors) {
        if (!pluginModule.preprocessors.oas3 && !pluginModule.preprocessors.oas2) {
          throw new Error(
            `Plugin \`preprocessors\` must have \`oas3\` or \`oas2\` preprocessors "${p}.`,
          );
        }
        plugin.preprocessors = {};
        if (pluginModule.preprocessors.oas3) {
          plugin.preprocessors.oas3 = prefixRules(pluginModule.preprocessors.oas3, id);
        }
        if (pluginModule.preprocessors.oas2) {
          plugin.preprocessors.oas2 = prefixRules(pluginModule.preprocessors.oas2, id);
        }
      }

      if (pluginModule.decorators) {
        if (!pluginModule.decorators.oas3 && !pluginModule.decorators.oas2) {
          throw new Error(`Plugin \`decorators\` must have \`oas3\` or \`oas2\` decorators "${p}.`);
        }
        plugin.decorators = {};
        if (pluginModule.decorators.oas3) {
          plugin.decorators.oas3 = prefixRules(pluginModule.decorators.oas3, id);
        }
        if (pluginModule.decorators.oas2) {
          plugin.decorators.oas2 = prefixRules(pluginModule.decorators.oas2, id);
        }
      }

      return plugin;
    })
    .filter(notUndefined);
}

export async function resolveApis({
  rawConfig,
  configPath = '',
  resolver,
}: {
  rawConfig: RawConfig;
  configPath?: string;
  resolver?: BaseResolver;
}): Promise<Record<string, ResolvedApi>> {
  const { apis = {}, lint: lintConfig = {} } = rawConfig;
  let resolvedApis: Record<string, ResolvedApi> = {};
  for (const [apiName, apiContent] of Object.entries(apis || {})) {
    if (apiContent.lint?.extends?.some(isNotString)) {
      throw new Error(
        `Error configuration format not detected in extends value must contain strings`,
      );
    }
    const rawLintConfig = getMergedLintRawConfig(lintConfig, apiContent.lint);
    const apiLint = await resolveLint({
      lintConfig: rawLintConfig,
      configPath,
      resolver,
    });
    resolvedApis[apiName] = { ...apiContent, lint: apiLint };
  }
  return resolvedApis;
}

async function resolveAndMergeNestedLint(
  {
    lintConfig,
    configPath = '',
    resolver = new BaseResolver(),
  }: {
    lintConfig?: LintRawConfig;
    configPath?: string;
    resolver?: BaseResolver;
  },
  parentConfigPaths: string[] = [],
  extendPaths: string[] = [],
): Promise<ResolvedLintConfig> {
  if (parentConfigPaths.includes(configPath)) {
    throw new Error(`Circular dependency in config file: "${configPath}"`);
  }
  const plugins = getUniquePlugins(
    resolvePlugins([...(lintConfig?.plugins || []), defaultPlugin], configPath),
  );
  const pluginPaths = lintConfig?.plugins
    ?.filter(isString)
    .map((p) => path.resolve(path.dirname(configPath), p));

  const resolvedConfigPath = isAbsoluteUrl(configPath)
    ? configPath
    : configPath && path.resolve(configPath);

  const extendConfigs: ResolvedLintConfig[] = await Promise.all(
    lintConfig?.extends?.map(async (presetItem) => {
      if (!isAbsoluteUrl(presetItem) && !path.extname(presetItem)) {
        return resolvePreset(presetItem, plugins);
      }
      const pathItem = isAbsoluteUrl(presetItem)
        ? presetItem
        : isAbsoluteUrl(configPath)
        ? new URL(presetItem, configPath).href
        : path.resolve(path.dirname(configPath), presetItem);
      const extendedLintConfig = await loadExtendLintConfig(pathItem, resolver);
      return await resolveAndMergeNestedLint(
        {
          lintConfig: extendedLintConfig,
          configPath: pathItem,
          resolver: resolver,
        },
        [...parentConfigPaths, resolvedConfigPath],
        extendPaths,
      );
    }) || [],
  );

  const { plugins: mergedPlugins = [], ...lint } = mergeExtends([
    ...extendConfigs,
    {
      ...lintConfig,
      plugins,
      extends: undefined,
      extendPaths: [...parentConfigPaths, resolvedConfigPath],
      pluginPaths,
    },
  ]);

  return {
    ...lint,
    extendPaths: lint.extendPaths?.filter((path) => path && !isAbsoluteUrl(path)),
    plugins: getUniquePlugins(mergedPlugins),
    recommendedFallback: lintConfig?.recommendedFallback,
    doNotResolveExamples: lintConfig?.doNotResolveExamples,
  };
}

export async function resolveLint(
  lintOpts: {
    lintConfig?: LintRawConfig;
    configPath?: string;
    resolver?: BaseResolver;
  },
  parentConfigPaths: string[] = [],
  extendPaths: string[] = [],
): Promise<ResolvedLintConfig> {
  const resolvedLint = await resolveAndMergeNestedLint(lintOpts, parentConfigPaths, extendPaths);

  return {
    ...resolvedLint,
    rules: resolvedLint.rules && groupLintAssertionRules(resolvedLint.rules),
  };
}

export function resolvePreset(presetName: string, plugins: Plugin[]): ResolvedLintConfig {
  const { pluginId, configName } = parsePresetName(presetName);
  const plugin = plugins.find((p) => p.id === pluginId);
  if (!plugin) {
    throw new Error(`Invalid config ${red(presetName)}: plugin ${pluginId} is not included.`);
  }

  const preset = plugin.configs?.[configName]! as ResolvedLintConfig;
  if (!preset) {
    throw new Error(
      pluginId
        ? `Invalid config ${red(
            presetName,
          )}: plugin ${pluginId} doesn't export config with name ${configName}.`
        : `Invalid config ${red(presetName)}: there is no such built-in config.`,
    );
  }
  return preset;
}

async function loadExtendLintConfig(
  filePath: string,
  resolver: BaseResolver,
): Promise<LintRawConfig> {
  try {
    const fileSource = await resolver.loadExternalRef(filePath);
    const rawConfig = transformConfig(parseYaml(fileSource.body) as RawConfig);
    if (!rawConfig.lint) {
      throw new Error(`Lint configuration format not detected: "${filePath}"`);
    }

    return rawConfig.lint;
  } catch (error) {
    throw new Error(`Failed to load "${filePath}": ${error.message}`);
  }
}

function getMergedLintRawConfig(configLint: LintRawConfig, apiLint?: LintRawConfig) {
  const resultLint = {
    ...configLint,
    ...apiLint,
    rules: { ...configLint?.rules, ...apiLint?.rules },
    oas2Rules: { ...configLint?.oas2Rules, ...apiLint?.oas2Rules },
    oas3_0Rules: { ...configLint?.oas3_0Rules, ...apiLint?.oas3_0Rules },
    oas3_1Rules: { ...configLint?.oas3_1Rules, ...apiLint?.oas3_1Rules },
    preprocessors: { ...configLint?.preprocessors, ...apiLint?.preprocessors },
    oas2Preprocessors: { ...configLint?.oas2Preprocessors, ...apiLint?.oas2Preprocessors },
    oas3_0Preprocessors: { ...configLint?.oas3_0Preprocessors, ...apiLint?.oas3_0Preprocessors },
    oas3_1Preprocessors: { ...configLint?.oas3_1Preprocessors, ...apiLint?.oas3_1Preprocessors },
    decorators: { ...configLint?.decorators, ...apiLint?.decorators },
    oas2Decorators: { ...configLint?.oas2Decorators, ...apiLint?.oas2Decorators },
    oas3_0Decorators: { ...configLint?.oas3_0Decorators, ...apiLint?.oas3_0Decorators },
    oas3_1Decorators: { ...configLint?.oas3_1Decorators, ...apiLint?.oas3_1Decorators },
    recommendedFallback: apiLint?.extends ? false : configLint.recommendedFallback,
  };
  return resultLint;
}

function groupLintAssertionRules(
  rules: Record<string, RuleConfig> | undefined,
): Record<string, RuleConfig> | undefined {
  if (!rules) {
    return rules;
  }

  // Create a new record to avoid mutating original
  const transformedRules: Record<string, RuleConfig> = {};

  // Collect assertion rules
  const assertions = [];
  for (const [ruleKey, rule] of Object.entries(rules)) {
    if (ruleKey.startsWith('assert/') && typeof rule === 'object' && rule !== null) {
      const assertion = rule;
      assertions.push({
        ...assertion,
        assertionId: ruleKey.replace('assert/', ''),
      });
    } else {
      // If it's not an assertion, keep it as is
      transformedRules[ruleKey] = rule;
    }
  }
  if (assertions.length > 0) {
    transformedRules.assertions = assertions;
  }

  return transformedRules;
}
