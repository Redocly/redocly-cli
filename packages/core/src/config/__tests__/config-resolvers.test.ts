import { resolveLint, resolveApis, resolveConfig } from '../config-resolvers';
const path = require('path');

import type { LintRawConfig, RawConfig } from '../types';

const configPath = path.join(__dirname, 'fixtures/resolve-config/.redocly.yaml');
const baseLintConfig: LintRawConfig = {
  rules: {
    'operation-2xx-response': 'warn',
  },
};

const minimalLintPreset = resolveLint({
  lintConfig: { ...baseLintConfig, extends: ['minimal'] },
});

const recommendedLintPreset = resolveLint({
  lintConfig: { ...baseLintConfig, extends: ['recommended'] },
});

const removeAbsolutePath = (item: string) =>
  item.match(/^.*\/packages\/core\/src\/config\/__tests__\/fixtures\/(.*)$/)![1];

describe('resolveLint', () => {
  it('should return the config with no recommended', async () => {
    const lint = await resolveLint({ lintConfig: baseLintConfig });
    expect(lint.plugins?.length).toEqual(1);
    expect(lint.plugins?.[0].id).toEqual('');
    expect(lint.rules).toEqual({
      'operation-2xx-response': 'warn',
    });
  });

  it('should return the config with correct order by preset', async () => {
    expect(
      await resolveLint({
        lintConfig: { ...baseLintConfig, extends: ['minimal', 'recommended'] },
      }),
    ).toEqual(await recommendedLintPreset);
    expect(
      await resolveLint({
        lintConfig: { ...baseLintConfig, extends: ['recommended', 'minimal'] },
      }),
    ).toEqual(await minimalLintPreset);
  });

  it('should return the same lintConfig when extends is empty array', async () => {
    const configWithEmptyExtends = await resolveLint({
      lintConfig: { ...baseLintConfig, extends: [] },
    });
    expect(configWithEmptyExtends.plugins?.length).toEqual(1);
    expect(configWithEmptyExtends.plugins?.[0].id).toEqual('');
    expect(configWithEmptyExtends.rules).toEqual({
      'operation-2xx-response': 'warn',
    });
  });

  it('should resolve extends with local file config', async () => {
    const config = {
      ...baseLintConfig,
      extends: ['local-config.yaml'],
    };

    const { plugins, ...lint } = await resolveLint({
      lintConfig: config,
      configPath,
    });

    expect(lint?.rules?.['operation-2xx-response']).toEqual('warn');
    expect(plugins).toBeDefined();
    expect(plugins?.length).toBe(2);

    expect(lint.extendPaths!.map(removeAbsolutePath)).toEqual([
      'resolve-config/.redocly.yaml',
      'resolve-config/local-config.yaml',
      'resolve-config/.redocly.yaml',
    ]);
    expect(lint.pluginPaths!.map(removeAbsolutePath)).toEqual(['resolve-config/plugin.js']);

    expect(lint.rules).toEqual({
      'boolean-parameter-prefixes': 'error',
      'local/operation-id-not-test': 'error',
      'no-invalid-media-type-examples': 'error',
      'operation-2xx-response': 'warn',
      'operation-description': 'error',
      'path-http-verbs-order': 'error',
    });
  });

  // TODO: fix circular test
  it.skip('should throw circular error', () => {
    const config = {
      ...baseLintConfig,
      extends: ['local-config-with-circular.yaml'],
    };
    expect(() => {
      resolveLint({ lintConfig: config, configPath });
    }).toThrow('Circular dependency in config file');
  });

  it('should resolve extends with local file config witch contains path to nested config', async () => {
    const lintConfig = {
      extends: ['local-config-with-file.yaml'],
    };
    const { plugins, ...lint } = await resolveLint({
      lintConfig,
      configPath,
    });

    expect(lint?.rules?.['no-invalid-media-type-examples']).toEqual('warn');
    expect(lint?.rules?.['operation-4xx-response']).toEqual('off');
    expect(lint?.rules?.['operation-2xx-response']).toEqual('error');
    expect(plugins).toBeDefined();
    expect(plugins?.length).toBe(3);

    expect(lint.extendPaths!.map(removeAbsolutePath)).toEqual([
      'resolve-config/.redocly.yaml',
      'resolve-config/local-config-with-file.yaml',
      'resolve-config/api/nested-config.yaml',
      'resolve-config/.redocly.yaml',
    ]);
    expect(lint.pluginPaths!.map(removeAbsolutePath)).toEqual([
      'resolve-config/api/plugin.js',
      'resolve-config/plugin.js',
      'resolve-config/api/plugin.js',
    ]);

    delete lint.extendPaths;
    delete lint.pluginPaths;
    expect(lint).toMatchSnapshot();
  });

  it('should resolve extends with url file config witch contains path to nested config', async () => {
    const lintConfig = {
      // This points to ./fixtures/resolve-remote-configs/remote-config.yaml
      extends: [
        'https://raw.githubusercontent.com/Redocly/openapi-cli/master/packages/core/src/config/__tests__/fixtures/resolve-remote-configs/remote-config.yaml',
      ],
    };

    const { plugins, ...lint } = await resolveLint({
      lintConfig,
      configPath,
    });

    expect(lint?.rules?.['operation-4xx-response']).toEqual('error');
    expect(lint?.rules?.['operation-2xx-response']).toEqual('error');
    expect(Object.keys(lint.rules || {}).length).toBe(2);

    expect(lint.extendPaths!.map(removeAbsolutePath)).toEqual([
      'resolve-config/.redocly.yaml',
      'resolve-config/.redocly.yaml',
    ]);
    expect(lint.pluginPaths!.map(removeAbsolutePath)).toEqual([]);
  });
});

describe('resolveApis', () => {
  it('should resolve apis lintConfig and merge minimal extends', async () => {
    const rawConfig: RawConfig = {
      apis: {
        petstore: {
          root: 'some/path',
          lint: {},
        },
      },
      lint: {
        extends: ['minimal'],
      },
    };
    const apisResult = await resolveApis({ rawConfig });
    expect(apisResult['petstore'].lint).toEqual(await minimalLintPreset);
  });

  it('should not merge recommended extends by default by every level', async () => {
    const rawConfig: RawConfig = {
      apis: {
        petstore: {
          root: 'some/path',
          lint: {},
        },
      },
      lint: {},
    };

    const apisResult = await resolveApis({ rawConfig, configPath });

    expect(apisResult['petstore'].lint.extendPaths!.map(removeAbsolutePath)).toEqual([
      'resolve-config/.redocly.yaml',
    ]);
    expect(apisResult['petstore'].lint.pluginPaths!.map(removeAbsolutePath)).toEqual([]);

    expect(apisResult['petstore'].lint.rules).toEqual({});
    //@ts-ignore
    expect(apisResult['petstore'].lint.plugins.length).toEqual(1);
    //@ts-ignore
    expect(apisResult['petstore'].lint.plugins[0].id).toEqual('');
  });

  it('should resolve apis lintConfig when it contains file and not set recommended', async () => {
    const rawConfig: RawConfig = {
      apis: {
        petstore: {
          root: 'some/path',
          lint: {
            rules: {
              'operation-4xx-response': 'error',
            },
          },
        },
      },
      lint: {
        rules: {
          'operation-2xx-response': 'warn',
        },
      },
    };

    const apisResult = await resolveApis({ rawConfig, configPath });
    expect(apisResult['petstore'].lint.rules).toEqual({
      'operation-2xx-response': 'warn',
      'operation-4xx-response': 'error',
    });
    //@ts-ignore
    expect(apisResult['petstore'].lint.plugins.length).toEqual(1);
    //@ts-ignore
    expect(apisResult['petstore'].lint.plugins[0].id).toEqual('');

    expect(apisResult['petstore'].lint.extendPaths!.map(removeAbsolutePath)).toEqual([
      'resolve-config/.redocly.yaml',
    ]);
    expect(apisResult['petstore'].lint.pluginPaths!.map(removeAbsolutePath)).toEqual([]);
  });

  it('should resolve apis lintConfig when it contains file', async () => {
    const rawConfig: RawConfig = {
      apis: {
        petstore: {
          root: 'some/path',
          lint: {
            extends: ['local-config.yaml'],
            rules: {
              'operation-4xx-response': 'error',
            },
          },
        },
      },
      lint: {
        extends: ['minimal'],
        rules: {
          'operation-2xx-response': 'warn',
        },
      },
    };

    const apisResult = await resolveApis({ rawConfig, configPath });
    expect(apisResult['petstore'].lint.rules).toBeDefined();
    expect(apisResult['petstore'].lint.rules?.['operation-2xx-response']).toEqual('warn'); // think about prioritize in merge ???
    expect(apisResult['petstore'].lint.rules?.['operation-4xx-response']).toEqual('error');
    expect(apisResult['petstore'].lint.rules?.['local/operation-id-not-test']).toEqual('error');
    //@ts-ignore
    expect(apisResult['petstore'].lint.plugins.length).toEqual(2);

    expect(apisResult['petstore'].lint.extendPaths!.map(removeAbsolutePath)).toEqual([
      'resolve-config/.redocly.yaml',
      'resolve-config/local-config.yaml',
      'resolve-config/.redocly.yaml',
    ]);
    expect(apisResult['petstore'].lint.pluginPaths!.map(removeAbsolutePath)).toEqual([
      'resolve-config/plugin.js',
    ]);
  });
});

describe('resolveConfig', () => {
  it('should add recommended to top level by default', async () => {
    const rawConfig: RawConfig = {
      apis: {
        petstore: {
          root: 'some/path',
          lint: {
            rules: {
              'operation-4xx-response': 'error',
            },
          },
        },
      },
      lint: {
        rules: {
          'operation-2xx-response': 'warn',
        },
      },
    };

    const { apis } = await resolveConfig(rawConfig, configPath);
    //@ts-ignore
    expect(apis['petstore'].lint.plugins.length).toEqual(1);
    //@ts-ignore
    expect(apis['petstore'].lint.plugins[0].id).toEqual('');

    expect(apis['petstore'].lint.extendPaths!.map(removeAbsolutePath)).toEqual([
      'resolve-config/.redocly.yaml',
    ]);
    expect(apis['petstore'].lint.pluginPaths!.map(removeAbsolutePath)).toEqual([]);

    expect(apis['petstore'].lint.rules).toEqual({
      ...(await recommendedLintPreset).rules,
      'operation-2xx-response': 'warn',
      'operation-4xx-response': 'error',
    });
  });

  it('should not add recommended to top level by default when apis have extends file', async () => {
    const rawConfig: RawConfig = {
      apis: {
        petstore: {
          root: 'some/path',
          lint: {
            extends: ['local-config.yaml'],
            rules: {
              'operation-4xx-response': 'error',
            },
          },
        },
      },
      lint: {
        rules: {
          'operation-2xx-response': 'warn',
        },
      },
    };

    const { apis } = await resolveConfig(rawConfig, configPath);
    expect(apis['petstore'].lint.rules).toBeDefined();
    expect(Object.keys(apis['petstore'].lint.rules || {}).length).toEqual(7);
    expect(apis['petstore'].lint.rules?.['operation-2xx-response']).toEqual('warn');
    expect(apis['petstore'].lint.rules?.['operation-4xx-response']).toEqual('error');
    expect(apis['petstore'].lint.rules?.['operation-description']).toEqual('error'); // from extends file config
    //@ts-ignore
    expect(apis['petstore'].lint.plugins.length).toEqual(2);

    expect(apis['petstore'].lint.extendPaths!.map(removeAbsolutePath)).toEqual([
      'resolve-config/.redocly.yaml',
      'resolve-config/local-config.yaml',
      'resolve-config/.redocly.yaml',
    ]);
    expect(apis['petstore'].lint.pluginPaths!.map(removeAbsolutePath)).toEqual([
      'resolve-config/plugin.js',
    ]);

    expect(apis['petstore'].lint.recommendedFallback).toBe(false);
  });

  it('should ignore minimal from the root and read local file', async () => {
    const rawConfig: RawConfig = {
      apis: {
        petstore: {
          root: 'some/path',
          lint: {
            extends: ['recommended', 'local-config.yaml'],
            rules: {
              'operation-4xx-response': 'error',
            },
          },
        },
      },
      lint: {
        extends: ['minimal'],
        rules: {
          'operation-2xx-response': 'warn',
        },
      },
    };

    const { apis } = await resolveConfig(rawConfig, configPath);
    expect(apis['petstore'].lint.rules).toBeDefined();
    expect(apis['petstore'].lint.rules?.['operation-2xx-response']).toEqual('warn');
    expect(apis['petstore'].lint.rules?.['operation-4xx-response']).toEqual('error');
    expect(apis['petstore'].lint.rules?.['operation-description']).toEqual('error'); // from extends file config
    //@ts-ignore
    expect(apis['petstore'].lint.plugins.length).toEqual(2);
    //@ts-ignore
    delete apis['petstore'].lint.plugins;

    expect(apis['petstore'].lint.extendPaths!.map(removeAbsolutePath)).toEqual([
      'resolve-config/.redocly.yaml',
      'resolve-config/local-config.yaml',
      'resolve-config/.redocly.yaml',
    ]);
    expect(apis['petstore'].lint.pluginPaths!.map(removeAbsolutePath)).toEqual([
      'resolve-config/plugin.js',
    ]);

    delete apis['petstore'].lint.extendPaths;
    delete apis['petstore'].lint.pluginPaths;
    expect(apis['petstore'].lint).toMatchSnapshot();
  });
});
