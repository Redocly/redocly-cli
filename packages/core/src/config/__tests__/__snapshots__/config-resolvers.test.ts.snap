// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolveConfig should ignore minimal from the root and read local file 1`] = `
Object {
  "decorators": Object {},
  "doNotResolveExamples": undefined,
  "oas2Decorators": Object {},
  "oas2Preprocessors": Object {},
  "oas2Rules": Object {},
  "oas3_0Decorators": Object {},
  "oas3_0Preprocessors": Object {},
  "oas3_0Rules": Object {
    "no-empty-servers": "error",
    "no-example-value-and-externalValue": "error",
    "no-invalid-media-type-examples": "error",
    "no-server-example.com": "warn",
    "no-server-trailing-slash": "error",
    "no-servers-empty-enum": "error",
    "no-undefined-server-variable": "error",
    "no-unused-components": "warn",
  },
  "oas3_1Decorators": Object {},
  "oas3_1Preprocessors": Object {},
  "oas3_1Rules": Object {
    "no-empty-servers": "error",
    "no-example-value-and-externalValue": "error",
    "no-server-example.com": "warn",
    "no-server-trailing-slash": "error",
    "no-servers-empty-enum": "error",
    "no-undefined-server-variable": "error",
    "no-unused-components": "warn",
  },
  "preprocessors": Object {},
  "recommendedFallback": false,
  "rules": Object {
    "assertions": "warn",
    "boolean-parameter-prefixes": "error",
    "info-contact": "off",
    "info-description": "warn",
    "info-license": "warn",
    "info-license-url": "warn",
    "local/operation-id-not-test": "error",
    "no-ambiguous-paths": "warn",
    "no-enum-type-mismatch": "error",
    "no-identical-paths": "error",
    "no-invalid-media-type-examples": "error",
    "no-path-trailing-slash": "error",
    "no-unresolved-refs": "error",
    "operation-2xx-response": "warn",
    "operation-4xx-response": "error",
    "operation-description": "error",
    "operation-operationId": "warn",
    "operation-operationId-unique": "error",
    "operation-operationId-url-safe": "error",
    "operation-parameters-unique": "error",
    "operation-security-defined": "error",
    "operation-singular-tag": "off",
    "operation-summary": "error",
    "operation-tag-defined": "off",
    "parameter-description": "off",
    "path-declaration-must-exist": "error",
    "path-http-verbs-order": "error",
    "path-not-include-query": "error",
    "path-parameters-defined": "error",
    "paths-kebab-case": "off",
    "spec": "error",
    "tag-description": "warn",
    "tags-alphabetical": "off",
  },
}
`;

exports[`resolveLint should resolve extends with local file config witch contains path to nested config 1`] = `
Object {
  "decorators": Object {},
  "doNotResolveExamples": undefined,
  "oas2Decorators": Object {},
  "oas2Preprocessors": Object {},
  "oas2Rules": Object {},
  "oas3_0Decorators": Object {},
  "oas3_0Preprocessors": Object {},
  "oas3_0Rules": Object {
    "no-empty-servers": "error",
    "no-example-value-and-externalValue": "error",
    "no-invalid-media-type-examples": "warn",
    "no-server-example.com": "warn",
    "no-server-trailing-slash": "error",
    "no-servers-empty-enum": "error",
    "no-undefined-server-variable": "error",
    "no-unused-components": "warn",
  },
  "oas3_1Decorators": Object {},
  "oas3_1Preprocessors": Object {},
  "oas3_1Rules": Object {
    "no-empty-servers": "error",
    "no-example-value-and-externalValue": "error",
    "no-server-example.com": "warn",
    "no-server-trailing-slash": "error",
    "no-servers-empty-enum": "error",
    "no-undefined-server-variable": "error",
    "no-unused-components": "warn",
  },
  "preprocessors": Object {},
  "recommendedFallback": undefined,
  "rules": Object {
    "assertions": Array [
      Object {
        "assertionId": "path-item-get-defined",
        "defined": true,
        "message": "Every path item must have a GET operation.",
        "property": "get",
        "subject": "PathItem",
      },
      Object {
        "assertionId": "tag-description",
        "message": "Tag description must be at least 13 characters and end with a full stop.",
        "minLength": 13,
        "pattern": "/\\\\.$/",
        "property": "description",
        "severity": "error",
        "subject": "Tag",
      },
    ],
    "boolean-parameter-prefixes": "error",
    "info-contact": "off",
    "info-description": "warn",
    "info-license": "warn",
    "info-license-url": "warn",
    "local/operation-id-not-test": "error",
    "no-ambiguous-paths": "warn",
    "no-enum-type-mismatch": "error",
    "no-identical-paths": "error",
    "no-invalid-media-type-examples": "warn",
    "no-path-trailing-slash": "error",
    "no-unresolved-refs": "error",
    "operation-2xx-response": "error",
    "operation-4xx-response": "off",
    "operation-description": "off",
    "operation-operationId": "warn",
    "operation-operationId-unique": "error",
    "operation-operationId-url-safe": "error",
    "operation-parameters-unique": "error",
    "operation-security-defined": "error",
    "operation-singular-tag": "off",
    "operation-summary": "error",
    "operation-tag-defined": "off",
    "parameter-description": "off",
    "path-declaration-must-exist": "error",
    "path-not-include-query": "error",
    "path-parameters-defined": "error",
    "paths-kebab-case": "off",
    "spec": "error",
    "tag-description": "warn",
    "tags-alphabetical": "off",
  },
}
`;
